src/libpgmoneta/management.c

```diff
@@ -3409,3 +3445,301 @@ print_info_json(struct json* json)
+			}
+		}
+	}
+	else if (!strcmp(backup_id, "latest") || !strcmp(backup_id, "newest"))
+	{
+		for (int i = number_of_backups - 1; backup_index == -1 && i >= 0; i--)
+		{
+			if (backups[i] != NULL)
+			{
+				backup_index = i;
+			}
+		}
+	}
+	else
+	{
+		for (int i = 0; backup_index == -1 && i < number_of_backups; i++)
+		{
+			if (backups[i] != NULL && !strcmp(backups[i]->label, backup_id))
+			{
+				backup_index = i;
+			}
+		}
+	}
+
+	//backup not found
+	if (backup_index == -1)
+	{
+		pgmoneta_log_error("Verify: No identifier for %s/%s", config->servers[srv].name, backup_id);
+		goto error;
+	}
+
+	//backup not valid
+	if (backups[backup_index]->valid == VALID_FALSE)
+	{
+		pgmoneta_log_error("Verify: Backup is not valid");
+		goto error;
+	}
+
+
+	//check if backup_manifest exists and is a regular file
+	backup_manifest = pgmoneta_get_server_backup_identifier_data(srv, backups[backup_index]->label);
+	backup_manifest = pgmoneta_append(backup_manifest, "backup_manifest");
+
+	if (pgmoneta_exists(backup_manifest) == false)
+	{
+		pgmoneta_log_error("Verify: backup_manifest does not exist.");
+		free(backup_manifest);
+		goto error;
+	}
+
+	if (pgmoneta_is_file(backup_manifest) == false)
+	{
+		pgmoneta_log_error("Verify: backup_manifest is not a file.");
+		free(backup_manifest);
+		goto error;
+	}
+
+	pgmoneta_log_debug("Verify: backup_manifest file found - OK");
+
+	//reading the backup_manifest
+	struct json_reader* reader = NULL;
+
+	if (pgmoneta_json_reader_init(backup_manifest, &reader) == 1)
+	{
+		pgmoneta_log_error("Verify: could not initialize JSON reader");
+		free(reader);
+		free(backup_manifest);
+		goto error;
+	}
+	//we don't need backup_manifest string anymore. all relevant info will be accessed through the reader
+	free(backup_manifest);
+	backup_manifest = NULL;
+
+	//go to to first array element, which is called Files
+	char* key_path[1] = {"Files"};
+
+	if (pgmoneta_json_locate(reader, key_path, 1) != 0)
+	{
+		pgmoneta_log_error("Verify: could not locate array Files in backup_manifest");
+		free(reader);
+		goto error;
+	}
+
+	//loop
+	backup_label = backups[backup_index]->label;
+	char **output = NULL;
+
+	output = (char **) malloc(sizeof(char**) * 4);
+
+	//main write loop
+	while(pgmoneta_verify_data(reader, output, srv, backup_label, report_all) == 0)
+	{
+		pgmoneta_log_debug("Verify: inside pgmoneta_management_write_verify_data loop");
+		pgmoneta_log_debug("Output[0] is %s", output[0]);
+		pgmoneta_log_debug("Output[1] is %s", output[1]);
+		pgmoneta_log_debug("Output[2] is %s", output[2]);
+		pgmoneta_log_debug("Output[3] is %s", output[3]);
+		pgmoneta_management_write_int32(socket, 1);
+		write_string("pgmoneta_management_write_verify_data", socket,  output[0]);
+		write_string("pgmoneta_management_write_verify_data", socket,  output[1]);
+		write_string("pgmoneta_management_write_verify_data", socket,  output[2]);
+		write_string("pgmoneta_management_write_verify_data", socket,  output[3]);
```

Comment:
I don't see you sending should_read at the end of each loop.

Reply:
Yet you read should_read at the end of each loop on the other side.

Reply:
You're right about this, Haoran. At line 3722 I send an int to the cli-side, to allow it to read the "first batch" of strings. However, when debugging, I don't see it reading even at this first iteration of the loop. This is what I'm stuck on now.

But you're right, I will need to fix the end of the loop in order to prepare the cli-side to read the strings generated in the following iterations of the loop.

Reply:
OK, try making the message sequence align on both side. And see if at least the first int32 reaches the other side.